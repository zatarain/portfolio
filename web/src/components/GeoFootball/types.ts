export interface Station {
	id: number,
	name: string,
	slug: string,
	country: string,
	time_zone: string,
	latitude: number,
	longitude: number,
	location: string,
	info_en: string,
	info_es: string,
}

export interface Stadium {
	id: number,
	name: string,
	slug: string,
	team: string,
	latitude: number,
	longitude: number,
	location: string,
}
