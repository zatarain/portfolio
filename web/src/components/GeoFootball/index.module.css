.map {
	height: 81vh;
}

.popup {
	display: grid;
	grid-template-columns: 2fr 4fr;
}

.popup dt {
	display: inline-block;
	font-weight: bold;
	margin: 0;
	text-align: right;
}

.popup dd {
	display: inline-block;
	margin: 0 1ch;
}
