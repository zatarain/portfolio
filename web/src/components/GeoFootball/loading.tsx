import Image from 'next/image'

interface Properties {
	text: string,
}

const Loading = ({ text }: Properties) => {
	return (
		<>
			<Image src="https://i.gifer.com/VAyR.gif" alt={text} width={16} height={16} /> {text}
		</>
	)
}

export default Loading
