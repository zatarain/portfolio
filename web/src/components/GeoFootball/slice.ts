import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { AppState } from '#store'
import type { Station } from './types'

export interface GeoFootballState {
	latitude: number,
	longitude: number,
}

const initialState = {
	latitude: 0,
	longitude: 50,
} as GeoFootballState


const BASE_URL = process.env.API_URL || ''

async function GET(path: string) {
	const response = await fetch(`${BASE_URL}${path}`)
	return await response.json()
}

export async function getStationsByCountry(): Promise<object> {
	const stations = await GET('/stations') as Array<Station>
	return stations.reduce((clusters: any, station: Station) => {
		clusters[station.country] = clusters[station.country] || []
		clusters[station.country].push(station)
		return clusters
	}, {});
}

const slice = createSlice({
	name: 'responsive',
	initialState,
	reducers: {
		get(state) {
			return state
		},
	}
})

export default slice.reducer
