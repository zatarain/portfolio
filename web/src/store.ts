import { configureStore, ThunkAction, Action, createSlice } from '@reduxjs/toolkit'

export function makeStore() {
  const responsiveFlipper = createSlice({
    name: 'flipper',
    initialState: {
      responsive: false,
    },
    reducers: {
      flipResponsive: (state) => {
        state.responsive = !state.responsive
      }
    }
  })
  return configureStore({
    reducer: { flipper: responsiveFlipper.reducer },
  })
}

const store = makeStore()

export type AppState = ReturnType<typeof store.getState>

export type AppDispatch = typeof store.dispatch

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  Action<string>
>

export default store
