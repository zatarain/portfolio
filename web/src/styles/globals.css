@import url('https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css');
@import url('https://fonts.cdnfonts.com/css/cascadia-code');
@import url('react-responsive-carousel/lib/styles/carousel.min.css');
@import url('leaflet/dist/leaflet.css');


:root {
  --bg-page: #fafbfc;
  --bg-primary: #28b873;
  --bg-secondary: #151b28;
  --bg-action: #f40;
  --bg-action-hover: #f60;
  --link-secondary: #a5acba;
  --highlight-secondary: #5cdeac;
  --fg-secondary: #fff;
  --fg-menuitem-hover: #7fffd4;
  --ldn-dark: #29bc9b;
  --ldn-lighter: #aaffec;
  --ldn-light: #79ffe1;
  --ldn: #50e3c2;
  --heading-2: var(--ldn-dark);
  --heading-3: var(--ldn-dark);
  --headings-font: Cambria, Cochin, Georgia, Times, 'Times New Roman', serif;
}

@media screen {
  html, body {
    padding: 0;
    margin: 0;
    text-align: center;
    font-family: Inter, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen, Ubuntu,
      Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
    font-size: 14px;
    scroll-behavior: smooth;
  }

  main {
    padding: 0 3rem;
    margin: 3rem 0;
    font-size: 1.2rem;
    line-height: 1.75rem;
  }
}

* {
  box-sizing: border-box;
}

a:link, a:visited {
  color: var(--ldn-dark);
  text-decoration: none;
}

a:hover, a:hover:visited {
  color: var(--bg-action);
}

.carousel ul.control-dots {
	bottom: 4rem;
	z-index: 3;
}

.carousel ul.control-dots li {
  width: 0.5rem;
  height: 0.5rem;
  padding: 0.5rem;
}

.carousel figcaption {
  margin: 0;
  position: relative;
  bottom: 3rem;
  font-size: 1rem;
}

.leaflet-bottom,
.leaflet-control,
.leaflet-pane,
.leaflet-top {
  z-index: 0 !important;
}

@media screen and (max-width: 850px) {
  .carousel figure img {
    height: 50vh;
  }

  .carousel figcaption {
    bottom: 1rem;
  }

  .carousel ul.control-dots {
    bottom: 3rem;
  }
}

@media screen and (max-width: 539px) {
  .carousel figure img {
    height: 30vh;
  }

  .carousel figcaption {
    bottom: 3rem;
    display: none;
  }

  .carousel ul.control-dots {
    bottom: -1rem;
  }
}
