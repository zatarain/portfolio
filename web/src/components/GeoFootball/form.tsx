import { Marker, Popup, useMapEvents } from 'react-leaflet'
import { LeafletEvent, Icon, Marker as LeafletMarker } from 'leaflet'
import { useForm } from 'react-hook-form'
import { useState } from 'react'
import { Noto_Color_Emoji } from 'next/font/google'

import type { Station, GroupedStations } from './types'
import Loading from './loading'
import { flag, saveStation } from './slice'

import styles from './index.module.css'
import { useDispatch } from 'react-redux'

const emoji = Noto_Color_Emoji({ weight: '400', subsets: ['emoji'], preload: false })

const pin = new Icon({
	iconUrl: 'https://cdn-icons-png.flaticon.com/512/2684/2684860.png',
	iconSize: [24, 24],
	iconAnchor: [0, 24],
})

interface Properties {
	clusters: GroupedStations,
	setClusters: Function,
}


const MapForm = ({ clusters, setClusters }: Properties) => {
	const countries = Object.keys(clusters)

	const initialStation = {
		name: '',
		country: 'GB',
		time_zone: 'Europe/London',
		latitude: 51.478,
		longitude: 0,
		info_en: '',
	} as Station

	const [station, setStation] = useState(initialStation)
	const [marker, setMarker] = useState(null)

	const {
		register,
		handleSubmit,
		formState: { isSubmitting, errors },
		reset,
		setError,
	} = useForm({
		defaultValues: { ...initialStation },
	});

	useMapEvents({
		dblclick(event) {
			setStation({
				...station,
				latitude: event.latlng.lat,
				longitude: event.latlng.lng,
			})
			marker && (marker as LeafletMarker).openPopup()
		}
	})

	const save = async (data: object | undefined) => {
		const response = await saveStation({
			...data,
			latitude: station.latitude,
			longitude: station.longitude,
		} as Station)
		if (response.ok) {
			marker && (marker as LeafletMarker).closePopup()
			const record = await response.json() as Station
			setStation({ ...initialStation });
			const cluster = clusters[record.country] as Station[]

			setClusters({
				...clusters,
				[record.country]: [...cluster, record],
			})
			reset()
		} else if (response.status == 400) {
			const errors: { [field: string]: string } = await response.json()
			for (const [field, message] of Object.entries(errors)) {
				setError(field as keyof Station, { type: 'custom', message })
			}
		} else {
			const message = await response.text()
			console.error(`Unknown error: ${message}`)
		}

		return response
	}

	const afterAdd = (event: LeafletEvent) => setMarker(event.target)

	return (
		<Marker position={[station.latitude, station.longitude]} icon={pin} eventHandlers={{ add: afterAdd }}>
			<Popup>
				<form id="popup-add-form" className={`${styles.form}`} method="POST" onSubmit={handleSubmit(save)}>
					<h3>Add New Marker</h3>
					<label htmlFor="name">Name:</label>
					<input id="name" type="text" role="textbox" required {...register('name')} />
					<div className={styles.error}>{errors.name?.message}</div>
					<div className={styles.location}>
						<div>
							<label htmlFor="latitude">Latitude:</label>
							<input id="latitude" type="text" name="latitude" value={station.latitude.toFixed(5)} disabled />
						</div>
						<div>
							<label htmlFor="longitude">Longitude:</label>
							<input id="longitude" type="text" name="longitude" value={station.longitude.toFixed(5)} disabled />
						</div>
					</div>
					<div className={styles['country-time']}>
						<div className={styles.country}>
							<label htmlFor="country">Country:</label>
							<select id="country" className={emoji.className} {...register('country')}>
								{countries.map((country: string) =>
									<option className={emoji.className} key={country} value={country}>{flag(country)}</option>)}
							</select>
						</div>
						<div className={styles.time}>
							<label htmlFor="time-zone">Timezone:</label>
							<input id="time-zone" type="text" {...register('time_zone')} />
						</div>
					</div>
					<label htmlFor="info-en">Additional information:</label>
					<input id="info-en" type="text" {...register('info_en')} />
					<button type="submit" disabled={isSubmitting}>
						{isSubmitting ? <Loading text="Saving..." /> : 'Save'}
					</button>
				</form>
			</Popup>
		</Marker>
	)
}

MapForm.displayName = 'MapForm'

export default MapForm
