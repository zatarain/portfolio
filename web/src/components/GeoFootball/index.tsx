import type { Station } from './types'
import { MapContainer, TileLayer, Marker, Popup } from 'react-leaflet'
import styles from './index.module.css'

interface Properties {
	stations: Station[]
}

const GeoFootball = ({ stations }: Properties) => {
	// var map = L.map('map').setView([51.505, -0.09], 13);
	return (
		<MapContainer className={styles.map} center={[51.505, -0.09]} zoom={13} scrollWheelZoom={false}>
			<TileLayer
				attribution='&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors'
				url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
			/>
			<Marker position={[51.505, -0.09]}>
				<Popup>
					A pretty CSS3 popup. <br /> Easily customizable.
				</Popup>
			</Marker>
		</MapContainer>
	)
}

export default GeoFootball
