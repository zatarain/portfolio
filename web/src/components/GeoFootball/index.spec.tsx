import { Provider } from 'react-redux'
import store from '#store'
import { render } from '@testing-library/react'
import { jest } from '@jest/globals'
import GeoFootball from '.'
import type { GroupedStations, Station } from './types'

jest.mock('react-leaflet', () => {
	return {
		__esModule: true,
		MapContainer: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		TileLayer: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		Marker: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		Popup: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
		useMapEvents: jest.fn(),
	}
})

jest.mock('react-leaflet-cluster', () => {
	return {
		__esModule: true,
		default: jest.fn(({ children }) => {
			return (<>{children}</>)
		}),
	}
})

jest.mock('next/font/google', () => {
	return {
		__esModule: true,
		Noto_Color_Emoji: jest.fn(() => {
			return {
				className: 'my-font',
			}
		}),
	}
})


describe('<GeoFootball ...>...</GeoFootball>', () => {
	it('renders the component correctly', async () => {
		const stations = {
			GB: [
				{
					id: 1,
					name: 'Canary Wharf',
					slug: 'canary-wharf',
					country: 'GB',
					time_zone: 'Europe/London',
					latitude: 51.50361,
					longitude: -0.01861,
					info_en: 'London Underground Station',
				} as Station,
			],
		} as GroupedStations

		const { asFragment } = render(
			<Provider store={store}>
				<GeoFootball stationsByCountry={stations} />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})
})
