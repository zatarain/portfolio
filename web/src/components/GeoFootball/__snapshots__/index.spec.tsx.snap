// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GeoFootball ...>...</GeoFootball> renders the component correctly 1`] = `
<DocumentFragment>
  A pretty CSS3 popup. 
  <br />
   Easily customizable.
</DocumentFragment>
`;
